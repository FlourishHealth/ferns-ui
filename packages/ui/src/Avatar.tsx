/* eslint-disable no-console */
import {ImageManipulator, ImageResult, SaveFormat} from "expo-image-manipulator";
import {launchImageLibraryAsync} from "expo-image-picker";
import {LinearGradient} from "expo-linear-gradient";
import React, {FC, useState} from "react";
import {Image, Pressable, Text, View} from "react-native";

import {AvatarProps, CustomSvgProps} from "./Common";
import {Icon} from "./Icon";
import {MobileIcon, OfflineIcon, OnlineIcon, OutOfOfficeIcon} from "./icons";
import {isMobileDevice} from "./MediaQuery";
import {useTheme} from "./Theme";
import {Tooltip} from "./Tooltip";

const sizes = {
  xs: 28,
  sm: 38,
  md: 56,
  lg: 72,
  xl: 120,
};

const initialsFontSizes = {
  xs: 12,
  sm: 16,
  md: 24,
  lg: 32,
  xl: 60,
};

const iconSizeScale = {
  xs: 0.5,
  sm: 0.7,
  md: 0.9,
  lg: 1.1,
  xl: 1.5,
};

const sizeIconPadding = {
  xs: 12,
  sm: 10,
  md: 9,
  lg: 7,
  xl: 0,
};

export const Avatar: FC<AvatarProps> = ({
  name,
  hasBorder = false,
  size = "md",
  src,
  onChange,
  status,
  doNotDisturb = false,
}) => {
  const {theme} = useTheme();
  const [isImageLoaded, setIsImageLoaded] = useState(true);
  const avatarImageFormat = SaveFormat.PNG;
  const avatarImageDiameter = sizes[size];
  const showEditIcon = status === "imagePicker";

  const avatarRadius = avatarImageDiameter / 2;
  const computedInitials =
    (name.match(/(^\S\S?|\s\S)?/g) ?? []) // Use nullish coalescing to handle the case where match returns null
      .map((v) => v.trim())
      .join("")
      .match(/(^\S|\S$)?/g) ??
    [] // Use nullish coalescing to handle the case where match returns null
      .join("")
      .toLocaleUpperCase();
  const statusIcons: {
    [id: string]: {
      icon: (props: CustomSvgProps) => React.ReactElement;
      label: string;
    };
  } = {
    online: {icon: OnlineIcon, label: "Online"},
    offline: {icon: OfflineIcon, label: "Offline"},
    outOfOffice: {icon: OutOfOfficeIcon, label: "Out of Office"},
    activeMobile: {
      icon: MobileIcon,

      label: "Active on Mobile",
    },
  };

  if (showEditIcon && !onChange) {
    console.warn("Avatars with the status of 'imagePicker' should also have an onChange property.");
  }

  const handleImageError = (event: any) => {
    setIsImageLoaded(false);
    console.warn("Image load error: ", event);
  };

  const pickImage = async () => {
    // TODO: Add permission request to use camera to take a picture
    const result = await launchImageLibraryAsync({
      mediaTypes: "images",
      allowsEditing: true,
      base64: true,
    });

    if (!result.canceled && result.assets) {
      const resizedImage = await resizeAndFormatImage(result.assets[0].uri);
      // convert base64 to data uri
      resizedImage.uri = `data:image/${avatarImageFormat.toLowerCase()};base64,${resizedImage.base64}`;
      if (onChange) {
        onChange({avatarImageFormat, ...resizedImage});
      }
    }
  };

  const resizeAndFormatImage = async (imageUri: string): Promise<ImageResult> => {
    const imageContext = await ImageManipulator.manipulate(imageUri);
    const resizedImage = await imageContext.resize({
      height: avatarImageDiameter,
    });
    const renderedImage = await resizedImage.renderAsync();
    return await renderedImage.saveAsync({format: avatarImageFormat, base64: true});
  };

  const renderEditIcon = () => {
    if (size !== "xl") {
      console.error(`Avatar: "imagePicker" status is only supported for size "xl"`);
      return null;
    }

    return (
      <Pressable
        aria-role="button"
        style={{
          alignItems: "center",
          backgroundColor: "rgba(255,255,255,0.75)",
          borderRadius: avatarRadius,
          height: avatarImageDiameter,
          justifyContent: "center",
          position: "absolute",
          width: avatarImageDiameter,
          zIndex: 5,
        }}
        onPress={pickImage}
      >
        <Icon color="primary" iconName="pen-to-square" size="2xl" type="regular" />
        <Text
          style={{
            textAlign: "center",
            fontWeight: "bold",
            fontSize: 12,
            marginTop: 10,
          }}
        >
          Upload Image
        </Text>
      </Pressable>
    );
  };

  const renderStatusIcon = () => {
    if (!status || showEditIcon) {
      return null;
    }
    const {icon} = statusIcons[status];

    if (!icon) {
      console.warn(`Avatar: Invalid status ${status}`);
      return null;
    }

    return (
      <View
        style={{
          bottom: 0,
          position: "absolute",
          right: 0,
          zIndex: 5,
        }}
      >
        {icon({
          doNotDisturb,
          transform: [{scale: iconSizeScale[size]}],
        })}
      </View>
    );
  };

  let avatar = (
    <View
      accessibilityHint={showEditIcon ? "Opens file explorer" : "Avatar image"}
      aria-label={`${name}'s avatar`}
      aria-role="image"
      style={{height: avatarImageDiameter, position: "relative", width: avatarImageDiameter}}
    >
      <Pressable
        aria-role="button"
        style={{
          overflow: "hidden",
          position: "relative",
          borderRadius: 1,
          cursor: showEditIcon ? "pointer" : "auto",
        }}
      >
        {src && isImageLoaded ? (
          // TODO: Make our Image component rounding work so that we can use it for Avatar.
          // Currently it creates an unrounded box around the Image.
          <Image
            accessibilityIgnoresInvertColors
            source={{uri: src, cache: "force-cache"}}
            style={{
              borderRadius: avatarRadius,
              borderWidth: hasBorder && status !== "imagePicker" ? avatarImageDiameter * 0.04 : 0,
              borderColor: hasBorder ? "white" : "transparent",
              height: avatarImageDiameter,
              overflow: "hidden",
            }}
            onError={handleImageError}
          />
        ) : (
          <View
            style={{
              height: avatarImageDiameter,
              width: avatarImageDiameter,
              borderRadius: avatarRadius,
              borderWidth: hasBorder && status !== "imagePicker" ? avatarImageDiameter * 0.04 : 0,
              borderColor: hasBorder && status !== "imagePicker" ? "white" : "transparent",
              display: "flex",
              alignItems: "center",
              justifyContent: "center",
              backgroundColor: theme.surface.secondaryDark,
            }}
          >
            <Text
              style={{
                fontWeight: 500,
                fontSize: initialsFontSizes[size],
                color: theme.text.inverted,
              }}
            >
              {computedInitials}
            </Text>
          </View>
        )}
      </Pressable>
      {/* Needs to come after the image so it renders on top. */}
      {showEditIcon && renderEditIcon()}
    </View>
  );

  if (hasBorder && status !== "imagePicker") {
    const gradientDiameter = avatarImageDiameter * 1.1;
    const gradientStartColor = "#FFC947";
    const gradientEndColor = "#EA9095";
    // Start the first color in the top left corner and end the second color in the bottom
    // right corner.

    avatar = (
      <LinearGradient
        colors={[gradientStartColor, gradientEndColor]}
        end={{x: 1, y: 1}}
        start={{x: 0, y: 0}}
        style={{
          height: gradientDiameter,
          width: gradientDiameter,
          borderRadius: gradientDiameter / 2,
          alignItems: "center",
          justifyContent: "center",
        }}
      >
        {avatar}
      </LinearGradient>
    );
  }

  if (status) {
    // Need to wrap the tooltip so it doesn't expand to 100% width and render the tooltip off.
    // Don't show the tooltips on mobile because they intercept the edit avatar clicks.
    const widthPlusPadding = avatarImageDiameter + sizeIconPadding[size];

    avatar = (
      <View
        style={{
          width: widthPlusPadding,
          paddingRight: sizeIconPadding[size],
          paddingBottom: sizeIconPadding[size],
        }}
      >
        <Tooltip idealPosition="top" text={isMobileDevice() ? undefined : status}>
          {avatar}
        </Tooltip>
        {renderStatusIcon()}
      </View>
    );
  }

  return avatar;
};
