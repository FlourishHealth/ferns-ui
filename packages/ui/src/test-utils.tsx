import {render} from "@testing-library/react-native";
import React from "react";
import {ThemeProvider} from "./Theme";

export const renderWithTheme = (ui: React.ReactElement) => {
  return render(<ThemeProvider>{ui}</ThemeProvider>);
};

export const createCommonMocks = () => ({
  onChange: jest.fn(),
  onFocus: jest.fn(),
  onBlur: jest.fn(),
  onEnter: jest.fn(),
  onSubmitEditing: jest.fn(),
  onIconClick: jest.fn(),
});

export const setupComponentTest = () => {
  jest.useFakeTimers();
  return createCommonMocks();
};

export const teardownComponentTest = () => {
  jest.useRealTimers();
  jest.clearAllMocks();
};
