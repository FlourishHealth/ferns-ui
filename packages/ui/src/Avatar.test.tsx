import {act, fireEvent} from "@testing-library/react-native";
// import {ImageManipulator} from "expo-image-manipulator";
// Import the actual implementation for SaveFormat
import * as ImageManipulator from "expo-image-manipulator";
import React from "react";

import {Avatar} from "./Avatar";
import {renderWithTheme} from "./test-utils";

// Mock expo-image-picker
jest.mock("expo-image-picker", () => ({
  launchImageLibraryAsync: jest.fn(),
}));

// Mock expo-image-manipulator
const mockResize = jest.fn();
const mockRenderAsync = jest.fn();
const mockSaveAsync = jest.fn();

jest.mock("expo-image-manipulator", () => {
  const actual = jest.requireActual("expo-image-manipulator");
  return {
    ...actual,
    ImageManipulator: {
      manipulate: jest.fn().mockImplementation(() => ({
        resize: mockResize.mockImplementation(() => ({
          renderAsync: mockRenderAsync.mockResolvedValue({
            saveAsync: mockSaveAsync.mockResolvedValue({
              uri: "test-uri",
              base64: "test-base64",
            }),
          }),
        })),
      })),
    },
    SaveFormat: {
      PNG: "png",
      JPEG: "jpeg",
    },
  };
});

jest.mock("expo-image-picker", () => ({
  launchImageLibraryAsync: jest.fn().mockResolvedValue({
    canceled: false,
    assets: [
      {
        uri: "test-uri",
        width: 100,
        height: 100,
      },
    ],
  }),
}));

// Mock the LinearGradient component
jest.mock("expo-linear-gradient", () => ({
  LinearGradient: "LinearGradient",
}));

describe("Avatar", () => {
  const defaultProps = {
    name: "John Doe",
    src: "https://example.com/avatar.jpg",
    testID: "avatar",
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly with default props", () => {
    const {toJSON} = renderWithTheme(<Avatar {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders initials when no image is provided", () => {
    const {getByText} = renderWithTheme(<Avatar name="John Doe" testID="avatar" />);
    expect(getByText("JD")).toBeTruthy();
  });

  it("renders image when src is provided", () => {
    const {getByTestId} = renderWithTheme(<Avatar {...defaultProps} />);
    const image = getByTestId("avatar-image");
    expect(image).toBeTruthy();
  });

  it("shows initials when image fails to load", () => {
    const {getByText, getByTestId} = renderWithTheme(<Avatar {...defaultProps} name="John Doe" />);

    // Simulate image load error
    fireEvent(getByTestId("avatar-image"), "onError");

    expect(getByText("JD")).toBeTruthy();
  });

  it("applies correct size class", () => {
    const size = "lg";
    const {getByTestId} = renderWithTheme(<Avatar {...defaultProps} size={size} />);
    const avatar = getByTestId("avatar-image");
    // Check if the style contains the expected size
    expect(avatar.props.style).toMatchObject({
      height: 72, // lg size from the sizes object
    });
  });

  it("shows status indicator when status is provided", () => {
    const {getByTestId} = renderWithTheme(<Avatar {...defaultProps} status="online" />);
    expect(getByTestId("status-indicator")).toBeTruthy();
  });

  it("shows edit icon when status is imagePicker and size is xl", () => {
    const {getByText} = renderWithTheme(
      <Avatar {...defaultProps} size="xl" status="imagePicker" />
    );
    expect(getByText("Upload Image")).toBeTruthy();
  });

  it("calls onChange when edit icon is pressed", async () => {
    const mockOnChange = jest.fn();
    const {getByText} = renderWithTheme(
      <Avatar {...defaultProps} size="xl" status="imagePicker" onChange={mockOnChange} />
    );

    await act(async () => {
      fireEvent.press(getByText("Upload Image"));
    });

    // The onChange should be called with the processed image
    expect(mockOnChange).toHaveBeenCalledWith({
      avatarImageFormat: "png",
      uri: "data:image/png;base64,test-base64",
      base64: "test-base64",
    });
    expect(ImageManipulator.ImageManipulator.manipulate).toHaveBeenCalled();
    expect(mockResize).toHaveBeenCalled();
    expect(mockRenderAsync).toHaveBeenCalled();
    expect(mockSaveAsync).toHaveBeenCalledWith({
      format: "png",
      base64: true,
    });
  });

  it("applies border when hasBorder is true", () => {
    const {getByTestId} = renderWithTheme(<Avatar {...defaultProps} hasBorder />);
    const avatar = getByTestId("avatar-image");
    // Check if the style contains border properties
    expect(avatar.props.style).toMatchObject({
      borderWidth: expect.any(Number),
      borderColor: expect.any(String),
    });
  });

  it("shows warning when imagePicker status is used with non-xl size", () => {
    const consoleWarnSpy = jest.spyOn(console, "warn").mockImplementation(() => {});
    renderWithTheme(<Avatar {...defaultProps} size="lg" status="imagePicker" />);

    expect(consoleWarnSpy).toHaveBeenCalledWith(
      "Avatars with the status of 'imagePicker' should also have an onChange property."
    );
    consoleWarnSpy.mockRestore();
  });
});
