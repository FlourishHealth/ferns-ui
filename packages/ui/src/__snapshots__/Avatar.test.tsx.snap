// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders correctly with default props 1`] = `
<View
  accessibilityHint="Avatar image"
  aria-label="John Doe's avatar"
  aria-role="image"
  style={
    {
      "height": 56,
      "position": "relative",
      "width": 56,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    aria-role="button"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "borderRadius": 1,
        "cursor": "auto",
        "overflow": "hidden",
        "position": "relative",
      }
    }
  >
    <Image
      accessibilityIgnoresInvertColors={true}
      onError={[Function]}
      source={
        {
          "cache": "force-cache",
          "uri": "https://example.com/avatar.jpg",
        }
      }
      style={
        {
          "borderColor": "transparent",
          "borderRadius": 28,
          "borderWidth": 0,
          "height": 56,
          "overflow": "hidden",
        }
      }
      testID="avatar-image"
    />
  </View>
</View>
`;
